test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+      function run() {
+        if (threadContext.poolEvents.filter(ev => getEventType(ev) === event).length >= count) {
+          return resolve();
+        }
+
+        threadContext.poolEventsEventEmitter.once('poolEvent', run);
+      }
+      run();
+    });
+  }
+});
+
+export class ThreadContext {
+  pool: ConnectionPool;
+  threads: Map<any, Thread>;
+  connections: Map<string, Connection>;
+  orphans: Set<Connection>;
+  poolEvents = [];
+  poolEventsEventEmitter = new EventEmitter();
+  hostAddress;
+  supportedOperations;
+
+  constructor(hostAddress) {
+    this.threads = new Map();
+    this.connections = new Map();
+    this.orphans = new Set();
+    this.poolEvents = [];
+    this.poolEventsEventEmitter = new EventEmitter();
+    this.hostAddress = hostAddress;
+    this.supportedOperations = getTestOpDefinitions(this);
+  }
+
+  getThread(name) {
+    let thread = this.threads.get(name);
+    if (!thread) {
+      thread = new Thread(this.supportedOperations);
+      this.threads.set(name, thread);
+    }
+
+    return thread;
+  }
+
+  createPool(options) {
+    this.pool = new ConnectionPool({ ...options, hostAddress: this.hostAddress });
+    ALL_POOL_EVENTS.forEach(ev => {
+      this.pool.on(ev, x => {
+        this.poolEvents.push(x);
+        this.poolEventsEventEmitter.emit('poolEvent');
+      });
+    });
+  }
+
+  closePool() {
+    return new Promise(resolve => {
+      ALL_POOL_EVENTS.forEach(ev => this.pool.removeAllListeners(ev));
+      this.pool.close(resolve);
+    });
+  }
+
+  async tearDown() {
+    if (this.pool) {
+      await this.closePool();
+    }
+    const connectionsToDestroy = Array.from(this.orphans).concat(
+      Array.from(this.connections.values())
+    );
+    const promises = connectionsToDestroy.map(conn => {
+      return new Promise<void>((resolve, reject) =>
+        conn.destroy({ force: true }, err => {
+          if (err) return reject(err);
+          resolve();
+        })
+      );
+    });
+    await Promise.all(promises);
+    this.poolEventsEventEmitter.removeAllListeners();
+  }
+}
+
+export async function runCmapTest(test: cmapTest, threadContext: ThreadContext) {
+  expect(knownTestKeys).to.include.members(Object.keys(test));
+
+  const poolOptions = test.poolOptions || {};
+  const operations = test.operations;
+  const expectedError = test.error;
+  const expectedEvents = test.events || [];
+  const ignoreEvents = test.ignore || [];
+
+  let actualError;
+
+  const MAIN_THREAD_KEY = Symbol('Main Thread');
+  const mainThread = threadContext.getThread(MAIN_THREAD_KEY);
+  mainThread.start();
+
+  threadContext.createPool(poolOptions);
+
+  for (const idx in operations) {
+    const op = operations[idx];
+
+    const threadKey = op.name === 'checkOut' ? op.thread || MAIN_THREAD_KEY : MAIN_THREAD_KEY;
+    const thread = threadContext.getThread(threadKey);
+
+    thread.queue(op);
+  }
+
+  await mainThread.finish().catch(e => {
+    actualError = e;
+  });
```

Comment:
```suggestion
const actualError = await mainThread.finish().catch(error => error);
```
We could still consider our inline catch pattern and update the else block to assert the value is just not a decendent of error? Or finish could return a struct of `{result, error}` I'm just thinking it's easier to get type information when the code doesn't asign to scope variable, but this comment is conditional on other type improvements so depends on where we land.

Reply:
I don't have a preference, again, I left the actual logic of the test execution unchanged as much as possible to limit the scope of stuff that needs to be reviewed

Reply:
left this part unchanged for now, let me know if you have a strong preference here